crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2352,16 +2352,107 @@ impl<'db> TypeInferenceBuilder<'db> {
         let ast::ExprCompare {
             range: _,
             left,
-            ops: _,
+            ops,
             comparators,
         } = compare;
 
-        self.infer_expression(left);
-        // TODO actually handle ops and return correct type
-        for right in comparators.as_ref() {
-            self.infer_expression(right);
+        // https://docs.python.org/3/reference/expressions.html#comparisons
+        // > Formally, if `a, b, c, …, y, z` are expressions and `op1, op2, …, opN` are comparison
+        // > operators, then `a op1 b op2 c ... y opN z` is equivalent to a `op1 b and b op2 c and
+        // ... > y opN z`, except that each expression is evaluated at most once.
+        let mut inferred_ty: Option<Type<'db>> = None;
+        std::iter::once(left.as_ref())
+            .chain(comparators.as_ref().iter())
+            // Evaluate expressions before iterating through pairs with `windows`
+            .map(|expr| self.infer_expression(expr))
+            // Allocation here is used to access the method `windows` not implemented on
+            // iterators (crates, experimental APIs or manual impl might provide it)
+            .collect::<Vec<Type<'db>>>()
+            .windows(2)
```

Comment:
I haven't looked at the PR as a whole yet, but you might be able to use [`itertools::Itertools::tuple_windows`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.tuple_windows) here. The method can be a bit dangerous to use, since it clones the iterators -- but `Type` instances should be very cheap to clone, so I think that's _probably_ a trade worth making to avoid allocating the `Vec`. You might have to add `itertools` as a dependency of the crate

Reply:
I'll try this when I can, thank you!